lib/gooddata/models/data_source.rb

```diff
@@ -0,0 +1,630 @@
+      end
+
+      # Create data source from json
+      # Expected keys:
+      # - :name (mandatory)
+      # - :alias (optional)
+      # - :prefix (optional)
+      # - :connectionInfo (mandatory)
+      # - :client (mandatory)
+      def create(opts)
+        ds_name = opts[:name]
+        ds_alias = opts[:alias]
+        ds_prefix = opts[:prefix]
+        ds_connection_info = opts[:connectionInfo]
+
+        GoodData.logger.info "Creating data source '#{ds_name}'"
+        fail ArgumentError, 'Data source name has to be provided' if ds_name.nil? || ds_name.blank?
+        fail ArgumentError, 'Data source connection info has to be provided' if ds_connection_info.nil?
+        fail ArgumentError, 'Data source does not support' unless support_data_source(ds_connection_info)
+
+        json = {
+          'dataSource' => {
+            'name' => ds_name,
+            'connectionInfo' => ds_connection_info
+          }
+        }
+        json['dataSource']['alias'] = ds_alias if ds_alias
+        json['dataSource']['prefix'] = ds_prefix if ds_prefix
+
+        # Create data source
+        c = GoodData.get_client(opts)
+        res = c.post(DATA_SOURCES_URL, json)
+
+        # create the public facing object
+        c.create(DataSource, res)
+      end
+
+      def support_data_source(connection_info_json)
+        (connection_info_json[SNOWFLAKE] || connection_info_json[REDSHIFT] || connection_info_json[BIGQUERY] ||
+            connection_info_json[GENERIC] || connection_info_json[S3] || connection_info_json[ADS])
+      end
+    end
+
+    def save
+      validate
+      validate_connection_info
+      if saved?
+        update_obj_json = client.put(uri, to_update_payload)
+        @json = update_obj_json
+      else
+        res = client.post(DATA_SOURCES_URL, to_update_payload)
+        fail 'Unable to create new Data Source' if res.nil?
+
+        @json = res
+      end
+      @connection_info = build_connection_info
+      self
+    end
+
+    def delete
+      saved? ? client.delete(uri) : nil
+    end
+
+    def initialize(json)
+      super
+      @json = json
+      validate
+      @connection_info = build_connection_info
+    end
+
+    def saved?
+      !uri.blank?
+    end
+
+    def name
+      @json['dataSource']['name']
+    end
+
+    def name=(new_name)
+      @json['dataSource']['name'] = new_name
+    end
+
+    def alias
+      @json['dataSource']['alias']
+    end
+
+    def alias=(new_alias)
+      @json['dataSource']['alias'] = new_alias
+    end
+
+    def prefix
+      @json['dataSource']['prefix']
+    end
+
+    def prefix=(new_prefix)
+      @json['dataSource']['prefix'] = new_prefix
+    end
+
+    def uri
+      @json['dataSource']['links']['self'] if @json && @json['dataSource'] && @json['dataSource']['links']
+    end
+
+    def id
+      uri.split('/')[-1]
+    end
+
+    def is(type)
+      @json['dataSource']['connectionInfo'][type]
+    end
+
+    private
+
+    def build_connection_info
```

Comment:
Should we have connection_info to get the connection_info?

Reply:
We already have declare "attr_accessor :connection_info" so We can get or set connection_info.